src/lib/comms/sol-http-common.c

```diff
@@ -123,3 +120,453 @@ sol_http_escape_string(char **escaped, const char *value)
+            !_is_sub_delimin(*i)) {
+            return -EINVAL;
+        }
+    }
+
+    s.len = i - partial_uri;
+
+    *next = i;
+    SOL_DBG("Uri server: %.*s", SOL_STR_SLICE_PRINT(s));
+
+    if (server)
+        *server = s;
+
+    return 0;
+}
+
+static int
+_get_port(char *partial_uri, int *port, char **next)
+{
+    struct sol_str_slice port_slice = sol_str_slice_from_str("80");
+    char *i;
+    int r = 0;
+
+    //The server port is not required.
+    if (!*partial_uri || *partial_uri != ':')
+        goto exit;
+
+    port_slice.data = ++partial_uri;
+    for (i = partial_uri; *i && *i != '/' && *i != '?' && *i != '#'; i++) {
+        if (!isdigit(*i)) {
+            SOL_ERR("Found a non digit while parsing the port. Found: %c", *i);
+            return -EINVAL;
+        }
+    }
+    port_slice.len = i - port_slice.data;
+    *next = i;
+
+exit:
+    SOL_DBG("URI Port: %.*s", SOL_STR_SLICE_PRINT(port_slice));
+    if (port) {
+        *port = sol_util_strtol(port_slice.data, NULL, port_slice.len, 10);
+    }
+    return r;
+}
+
+static int
+_get_path(char *partial_uri, struct sol_str_slice *path, char **next)
+{
+    struct sol_str_slice p = SOL_STR_SLICE_EMPTY;
+    char *i = partial_uri;
+
+    if (!*i)
+        goto exit;
+
+    if (*i != '/')
+        goto exit;
+
+    i++;
+
+    //Oops, this is not a path.
+    if (*i == '?')
+        goto exit;
+
+    p.data = i;
+    for (; *i && *i != '?' && *i != '#'; i++) {
+        if (*i != '/' && !_is_pchar(*i)) {
+            SOL_ERR("Invalid character while parsing the path. Found: %c", *i);
+            return -EINVAL;
+        }
+    }
+    p.len = i - p.data;
+
+exit:
+    SOL_DBG("URI Path: *%.*s*", SOL_STR_SLICE_PRINT(p));
+    *next = i;
+    if (path)
+        *path = p;
+    return 0;
+}
+
+static int
+_get_query(char *partial_uri, struct sol_http_param *query_params, char **next)
+{
+    struct sol_str_slice query, *token;
+    char *i = partial_uri, *sep;
+    uint16_t j;
+    struct sol_vector tokens;
+    int r = 0;
+
+    if (!*i)
+        goto exit;
+
+    if (*i != '?')
+        goto exit;
+
+    query.data = ++i;
+
+    for (; *i && *i != '#'; i++) {
+        if (*i != '/' && *i != '?' && !_is_pchar(*i)) {
+            SOL_ERR("Invalid character while parsing the query. "
+                "Found: %c", *i);
+            return -EINVAL;
+        }
+    }
+    query.len = i - query.data;
+
+    tokens = sol_util_str_split(query, "&", 0);
+
+    SOL_VECTOR_FOREACH_IDX (&tokens, token, j) {
+        struct sol_http_param_value param;
+        sep = memchr(token->data, '=', token->len);
+        if (!sep) {
+            SOL_ERR("The param: %.*s has no \"=\" separator",
+                SOL_STR_SLICE_PRINT(*token));
+            r = -EINVAL;
+            break;
+        }
+
+        param.value.key_value.key.data = token->data;
+        param.value.key_value.key.len = sep - token->data;
+
+        param.value.key_value.value.data = sep + 1;
```

Comment:
you're not unescaping value, this is problematic.

However, doing so would allocate memory. To solve it I'd not split the query, only return a slice with the query contents and point the query split function, it would allocate memory and point to free function (clears vector and values/keys).